dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java

```diff
@@ -97,6 +104,47 @@ public void validateTrackedEntity( ValidationErrorReporter reporter, TrackedEnti
         OrganisationUnit organisationUnit = context.getOrganisationUnit( trackedEntity.getOrgUnit() );
 
         validateAttributes( reporter, trackedEntity, tei, organisationUnit );
+        validateMandatoryAttributes( reporter, trackedEntity );
+    }
+
+    private void validateMandatoryAttributes( ValidationErrorReporter reporter, TrackedEntity trackedEntity )
+    {
+        TrackedEntityType trackedEntityType = reporter.getValidationContext()
+            .getTrackedEntityType( trackedEntity.getTrackedEntityType() );
+
+        if ( trackedEntityType != null )
+        {
+            Set<String> trackedEntityAttributes = Optional.of( trackedEntity )
+                .map( TrackedEntity::getAttributes )
+                .orElse( Collections.emptyList() )
+                .stream()
+                .map( Attribute::getAttribute )
+                .collect( Collectors.toSet() );
+
+            trackedEntityType.getTrackedEntityTypeAttributes()
+                .stream()
+                .filter( this::isMandatory )
```

Comment:
I found difficult following this Optional.of and then orElse flow and I am not sure if in this case you could really have a null TrackedEntityTypeAttribute in the list of trackedEntityType.getTrackedEntityTypeAttributes()
I would go with 

```suggestion
               .filter(Obects::isNotNull)
               .filter(TrackedEntityTypeAttribute::isMandatory)
```

but of course this is purely personal style

Reply:
yes, your version is much easier, I'll change it